### PR TITLE
Drawing: add drawing tool subtasks to the models

### LINKED ISSUE #1239: RFC: drawing sub-task
**Package**
lib-classifier

**Is your feature request related to a problem? Please describe.**
https://github.com/zooniverse/front-end-monorepo/blob/1edd001edb31550a023f51b1aa15bbe70e675b3b/packages/lib-classifier/src/store/tasks/Task.js#L7-L26
Task models get their annotations from the classifications store and call `classifications.addAnnotation(value, task)` on update. It should be possible to change both of those depending on the task context (workflow task or drawing subtask.)

**Describe the solution you'd like**
Tasks should be agnostic about where their annotations come from. I think the task view component could have `annotation` and `onComplete` (or `onChange`?) as props. The task area would pass in the annotation from the classification, and `classifications.addAnnotation` as the callback. Drawing subtasks would pass in the annotation from the drawing mark, and a callback that updates the drawing tool annotation.

There might be better solutions which involve passing context to the task model instead, so I'm opening this issue for discussion. Is there a way to replace `getRoot(self).classifications` with something more flexible in the Task model code?

### PR DESCRIPTION
**Pull request by eatyourgreens**
#1366 redone. This time I remembered to add `tool.details` to drawing tools, and wrote the spec for task and annotation behaviour first.

- add a task to a drawing tool by calling `tool.createTask(snapshot)`.
- access subtasks for a mark via `mark.tasks`.
- check whether the subtasks have been completed for a mark by checking `mark.isComplete`.
- use `afterCreate` in a drawing task to generate tasks from tool details.

TODO:
- [x] Block the Next button until all required subtasks have been annotated.

Package:
lib-classifier

Closes #1239.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

**Comment by srallen** [comment]
I'm going to write an ADR for sub-tasks to take what's been discussed across a couple of different issues, including #500, so we can use that for reviewing this PR.

File: packages/lib-classifier/src/plugins/drawingTools/models/marks/Mark/Mark.js [hunk 1]
```diff
@@ -16,12 +16,20 @@ const BaseMark = types.model('BaseMark', {
   toolIndex: types.optional(types.number, 0)
 })
   .views(self => ({
+    get isComplete () {
+      return self.tasks.reduce((isMarkComplete, task) => !task.required || self.annotation(task).isComplete, true)
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/plugins/drawingTools/models/marks/Mark/Mark.js]: `isMarkComplete` is missing from the RHS of the reducer.

**Comment by srallen** [comment]
I've rebased the branch and fixed the subsequent broken tests.

**Comment by eatyourgreens** [comment]
This PR isn’t intended to handle how marks are mapped to `classification.annotations` for classification submission. I’d prefer changes for that to be their own PR. 

`markIndex` won’t be known until the marks map is converted to an array on submission, so I’d prefer that to be calculated on classification completion. If we add an array index to each map item here, we’ll have to maintain those indexes for all marks while drawing is in progress. That’s going to mean updating the whole map whenever we remove a mark, unless I’m missing something.

**Comment by eatyourgreens** [comment]
The tools Readme already has the API changes from this PR. I've made a couple of small changes.

**Comment by srallen** [comment]
> markIndex won’t be known until the marks map is converted to an array on submission, so I’d prefer that to be calculated on classification completion. If we add an array index to each map item here, we’ll have to maintain those indexes for all marks while drawing is in progress. That’s going to mean updating the whole map whenever we remove a mark, unless I’m missing something.

That's all true, but that also doesn't mean that it can't be added to the model as `markIndex: types.maybe(types.number))` in this PR.

**Comment by eatyourgreens** [comment]
>That's all true, but that also doesn't mean that it can't be added to the model as markIndex: types.maybe(types.number)) in this PR.

It will always be undefined, even for subtasks. That's why I'm questioning it. `markIndex` doesn't have a value until we build a modified copy of the classification for submission to Panoptes, at which point we aren't working with the models any more.

I'd definitely be more comfortable modifying the annotation models in a separate PR.

**Comment by eatyourgreens** [comment]
#1443, which is only half-written at the moment.

File: packages/lib-classifier/src/plugins/drawingTools/README.md [hunk 2]
```diff
@@ -69,3 +70,64 @@ In addition, mark models should extend the base Mark model with any properties s
...
+  instruction: 'Transcribe the marked line.',
+  required: true,
+  type: 'text'
+})
+
+// draw some lines
+
+const line1 = tool.createMark({
+  id: 'line1'
+  x1: 10,
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/drawingTools/README.md]: `tool.createMark()` adds the correct `toolType` to the snapshot before creating the mark.

File: packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js [hunk 3]
```diff
@@ -25,12 +36,23 @@ const Tool = types.model('Tool', {
       return newMark
     }
 
+    function createTask (snapshot) {
+      try {
+        self.tasks.push(snapshot)
+        return self.tasks[self.tasks.length - 1]
```
[the remark this answers is not in the document]
Comment (eatyourgreens) [reply packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js]: So we can get the new task with `const task = tool.createTask(snapshot)`.

**Pull request merged**